plugins/producer_plugin/include/eosio/producer_plugin/block_timing_util.hpp

```diff
@@ -25,22 +52,95 @@ namespace block_timing_util {
-            // production round. In this case, we just use the hard deadline.
-            return std::min(hard_deadline, now + fc::microseconds(cpu_effort_us));
+   inline fc::time_point calculate_producing_block_deadline(uint32_t cpu_effort_us, chain::block_timestamp_type block_time) {
+      auto estimated_deadline = production_round_block_start_time(cpu_effort_us, block_time) + fc::microseconds(cpu_effort_us);
+      auto now                = fc::time_point::now();
+      if (estimated_deadline > now) {
+         return estimated_deadline;
+      } else {
+         // This could only happen when the producer stop producing and then comes back alive in the middle of its own
+         // production round. In this case, we just use the hard deadline.
+         const auto hard_deadline = block_time.to_time_point() - fc::microseconds(chain::config::block_interval_us - cpu_effort_us);
+         return std::min(hard_deadline, now + fc::microseconds(cpu_effort_us));
+      }
+   }
+
+   inline uint32_t calculate_next_block_slot(const chain::account_name& producer_name, uint32_t current_block_slot, uint32_t block_num,
+                                             const std::vector<chain::producer_authority>& active_schedule, const producer_watermarks& prod_watermarks) {
+      // determine if this producer is in the active schedule and if so, where
+      auto itr =
+         std::find_if(active_schedule.begin(), active_schedule.end(), [&](const auto& asp) { return asp.producer_name == producer_name; });
+      if (itr == active_schedule.end()) {
+         // this producer is not in the active producer set
+         return UINT32_MAX;
+      }
+
+      size_t   producer_index = itr - active_schedule.begin();
+      uint32_t minimum_offset = 1; // must at least be the "next" block
+
+      // account for a watermark in the future which is disqualifying this producer for now
+      // this is conservative assuming no blocks are dropped.  If blocks are dropped the watermark will
+      // disqualify this producer for longer but it is assumed they will wake up, determine that they
+      // are disqualified for longer due to skipped blocks and re-calculate their next block with better
+      // information then
+      auto current_watermark = prod_watermarks.get_watermark(producer_name);
+      if (current_watermark) {
+         const auto watermark = *current_watermark;
+         if (watermark.first > block_num) {
+            // if I have a watermark block number then I need to wait until after that watermark
+            minimum_offset = watermark.first - block_num + 1;
          }
+         if (watermark.second.slot > current_block_slot) {
+            // if I have a watermark block timestamp then I need to wait until after that watermark timestamp
+            minimum_offset = std::max(minimum_offset, watermark.second.slot - current_block_slot + 1);
```

Comment:
Seems inconsistent with line 91? Shouldn't it be:
```suggestion
            minimum_offset = std::max(minimum_offset, watermark.second.slot - block_num + 1);
```

BTW, I don't  understand the difference between `chain.head_block_num()` and `chain.head_block_time().slot`?

Reply:
No. 91 is working with a block_num watermark. Line 95 is working with a block_slot watermark.

Block numbers can never be skipped, but blocks can fail to be produced. When blocks are not produced when they should be block slots are missed. Block slots are just an abbreviated time. 

To illustrate this check out `to_time_point()` of `block_timestamp`:
```
            int64_t msec = slot * (int64_t)IntervalMs;
            msec += EpochMs;
            return fc::time_point(fc::milliseconds(msec));
```
```
Example:

block_slot = 13 - time 2023-08-11T10:00:00.000 - block_num 50
block_slot = 14 - time 2023-08-11T10:00:00.500 - block_num 51
block_slot = 15 - time 2023-08-11T10:00:01.000 - block_num 52
block_slot = 16 - time 2023-08-11T10:00:01.500 - block fails to produce
block_slot = 17 - time 2023-08-11T10:00:02.000 - block fails to produce
block_slot = 18 - time 2023-08-11T10:00:02.500 - block_num 53
```

Reply:
Thanks for the explanation Kevin!